src/Chainweb/Pact/PactService.hs

```diff
@@ -489,17 +489,25 @@ validateChainwebTxs
     -> IO (Vector Bool)
 validateChainwebTxs dbEnv cp blockOriginationTime bh txs
     | bh == 0 = pure $! V.replicate (V.length txs) True
-    | otherwise = balances txs >>= newIORef >>= \bsr -> V.mapM (validate bsr) txs
+    | otherwise = do
+          let f t = let p = view P.cmdHash t
+                    in (not . isJust) <$> _cpLookupProcessedTx cp p
+          dupecheckOks <- V.mapM f txs
+          let txs' = V.zip txs dupecheckOks
+          balances txs' >>= newIORef >>= \bsr -> V.mapM (validate bsr) txs'
   where
-    validate :: IORef Balances -> ChainwebTransaction -> IO Bool
-    validate bsr tx = do
-        bs <- readIORef bsr
-        case HM.lookup sender bs >>= debitGas bs tx of
-            Nothing -> pure False
-            Just bs' -> do
-                let !valid = all ($ tx) validations
-                when valid $ writeIORef bsr bs'
-                pure valid
+    validate :: IORef Balances -> (ChainwebTransaction, Bool) -> IO Bool
```

Comment:
Does returning just `Bool` provide sufficient information in case of an error? It seems that there are different things validated within this function (dup-checks and balance checks).

Reply:
Colin and I talked about this. Return type should probably be `Either Mempool.InsertError ()` but this is ok for now